safenode/src/protocol/types/wallet/local_store.rs

```diff
@@ -0,0 +1,344 @@
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use super::{CreatedDbc, DepositWallet, Result, SendClient, SendWallet};
+
+use crate::protocol::types::transfers::Outputs as TransferDetails;
+
+use sn_dbc::{Dbc, DbcId, DbcIdSource, MainKey, PublicAddress, Token};
+
+use async_trait::async_trait;
+use std::collections::{BTreeMap, BTreeSet};
+
+/// A wallet that can send tokens to other addresses.
+pub struct LocalSendWallet<C: SendClient> {
+    client: C,
+    wallet: LocalDepositWallet,
+}
+
+/// A wallet that can only receive tokens.
+pub struct LocalDepositWallet {
+    /// The secret key with which we can access
+    /// all the tokens in the available_dbcs.
+    key: MainKey,
+    /// The current balance of the wallet.
+    balance: Token,
+    /// These are dbcs we've owned, that have been
+    /// spent when sending tokens to other addresses.
+    spent_dbcs: BTreeMap<DbcId, Dbc>,
+    /// These are the dbcs we own that are not yet spent.
+    available_dbcs: BTreeMap<DbcId, Dbc>,
+    /// These are the dbcs we've created by
+    /// sending tokens to other addresses.
+    output_history: Vec<CreatedDbc>,
+    // /// The path to the wallet file.
+    // local_store: std::path::PathBuf,
+}
+
+impl DepositWallet for LocalDepositWallet {
+    fn new(key: MainKey) -> Self {
+        Self {
+            key,
+            balance: Token::zero(),
+            spent_dbcs: BTreeMap::new(),
+            available_dbcs: BTreeMap::new(),
+            output_history: vec![],
+        }
+    }
+
+    // /// Loads a serialized wallet from a path.
+    // fn load_from(path: &Path) -> Self {
+    //     Self {
+    //     }
+    // }
+
+    fn address(&self) -> PublicAddress {
+        self.key.public_address()
+    }
+
+    fn balance(&self) -> Token {
+        self.balance
+    }
+
+    fn new_dbc_address(&self) -> DbcIdSource {
+        self.key.random_dbc_id_src(&mut rand::thread_rng())
+    }
+
+    fn deposit(&mut self, dbcs: Vec<Dbc>) {
+        if dbcs.is_empty() {
+            return;
+        }
+
+        let mut received_dbcs = dbcs
+            .into_iter()
+            .filter_map(|dbc| {
+                let id = dbc.id();
+                (!self.spent_dbcs.contains_key(&id)).then_some((id, dbc))
+            })
+            .filter_map(|(id, dbc)| dbc.derived_key(&self.key).is_ok().then_some((id, dbc)))
+            .collect();
+
+        self.available_dbcs.append(&mut received_dbcs);
+
+        let new_balance = self
+            .available_dbcs
+            .iter()
+            .flat_map(|(_, dbc)| {
+                dbc.derived_key(&self.key)
+                    .map(|derived_key| (dbc, derived_key))
+            })
+            .flat_map(|(dbc, derived_key)| dbc.revealed_input(&derived_key))
+            .fold(0, |total, amount| total + amount.revealed_amount().value());
+
+        self.balance = Token::from_nano(new_balance);
```

Comment:
why not just calculate a total amount from `received_dbcs` then:
`self.balance += total_balance_from_received_dbc` ?
which shall save certain computations?

Reply:
That's a way it could be done as well. Just consider the other approach slightly more solid, to not rely on intermediate state. And the computation cost is negligible.